gordon_janitor_gcp/gcp_clients.py

```diff
@@ -0,0 +1,234 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Client classes to retrieve project and instance data from GCE.
+
+These clients use the asynchronous HTTP client defined in
+:py:mod:`gordon_janitor_gcp.http_client` and require service account or
+JWT-token credentials for authentication.
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    import aiohttp
+
+    from gordon_janitor_gcp import auth
+
+    loop = asyncio.get_event_loop()
+
+    async def main():
+        session = aiohttp.ClientSession()
+        auth_client = auth.GoogleAuthClient(
+            keyfile='/path/to/keyfile', session=session)
+        client = GCEClient(auth_client, session)
+        instances = await client.list_instances('project-id')
+        print(instances)
+
+    loop.run_until_complete(main())
+    # example output
+    # [{'hostname': 'instance-1', 'internal_ip': '10.10.10.10',
+    #   'external_ip': '192.168.1.10'}]
+"""
+
+import logging
+
+from gordon_janitor_gcp import http_client
+
+
+class GCRMClient(http_client.AIOGoogleHTTPClient,
+                 http_client.GPaginatorMixin):
+    """Async client to interact with Google Cloud Resource Manager API.
+
+    You can find the endpoint documentation
+    `here <https://cloud.google.com/resource-manager/
+    reference/rest/#rest-resource-v1projects>`__.
+
+    Attributes:
+        BASE_URL (str): Base endpoint URL.
+
+    Args:
+        auth_client (gordon_janitor_gcp.auth.GoogleAuthClient): client
+            to manage authentication for HTTP API requests.
+        session (aiohttp.ClientSession): (optional) ``aiohttp`` HTTP
+            session to use for sending requests. Defaults to the
+            session object attached to ``auth_client`` if not provided.
+    """
+    BASE_URL = 'https://cloudresourcemanager.googleapis.com'
+
+    def __init__(self, auth_client=None, session=None):
+        super().__init__(auth_client=auth_client, session=session)
+
+    def _parse_rsps_for_projects(self, responses):
+        projects = []
+        for response in responses:
+            for project in response.get('projects', []):
+                projects.append(project)
+        return projects
+
+    async def list_all_active_projects(self, api_version='v1', page_size=1000):
```

Comment:
Is there a reason you have the API version here, and not given in the `__init__`?

Reply:
Also, don't forget docstrings for this.

Reply:
> Is there a reason you have the API version here, and not given in the __init__?

Is there a benefit of having it in the __init__ vs. here?

Reply:
I guess I see it weird that there's a class attribute of the base URL, but the "completion" of that URL is in a method rather than class instantiation. Do you anticipate needing to make calls to the same endpoint but the different version?

Reply:
Ah, gotcha! Thanks for explaining it!